lib/percy.rb

```diff
@@ -0,0 +1,119 @@
+require 'logger'
+require 'net/http'
+require 'uri'
+require 'json'
+require 'environment'
+
+module Percy
+  # Takes a snapshot of the given page HTML and its assets.
+  #
+  # See https://docs.percy.io/v1/docs/configuration for detailed documentation on
+  # snapshot options.
+  #
+  # @param [Capybara::Session] page The Capybara page to snapshot.
+  # @param [Hash] options
+  # @option options [String] :name A unique name for the current page that identifies
+  #   it across builds. By default this is the URL of the page, but can be customized if the
+  #   URL does not entirely identify the current state.
+  # @option options [Array(Number)] :widths Widths, in pixels, that you'd like to capture for
+  #   this snapshot.
+  def self.snapshot(page, options = {})
+    return unless self._is_agent_running?
+
+    domSnapshot = self._make_dom_snapshot(page)
+    return unless domSnapshot
+
+    if !options.has_key?(:name)
+      options[:name] = page.current_url
+    end
+
+    body = {
+      url: page.current_url,
+      domSnapshot: domSnapshot,
+      clientInfo: Percy.client_info,
+      environmentInfo: Percy.environment_info,
+    }
+    body = body.merge(self._keys_to_json(options))
+    self._post_snapshot_to_agent(body)
+  end
+
+  private
+
+  AGENT_HOST = 'localhost'
+  # Technically, the port is configurable when you run the agent. One day we might want
```

Comment:
We could make `percy exec` set an environment variable with this port before it starts the subprocess, so subprocesses could pick this up dynamically.

(Probably not required for this PR)